pkg/util/kubernetes/apiserver/hpa_controler.go

```diff
@@ -0,0 +1,312 @@
+				c.toStore.m.Unlock()
+				if !c.le.IsLeader() || len(data) == 0 {
+					continue
+				}
+
+				log.Tracef("Batch call pushing %d metrics", len(data))
+				err := c.pushToGlobalStore(data)
+				if err != nil {
+					log.Errorf("Error storing the list of External Metrics to the ConfigMap: %v", err)
+					continue
+				}
+			}
+		}
+	}()
+}
+
+func (h *AutoscalersController) pushToGlobalStore(toPush []custommetrics.ExternalMetricValue) error {
+	// reset the batch before submitting to avoid a discrepancy between the global store and the local one
+	h.toStore.m.Lock()
+	h.toStore.data = nil
+	h.toStore.m.Unlock()
+	return h.store.SetExternalMetricValues(toPush)
+}
+
+func (h *AutoscalersController) updateExternalMetrics() {
+	emList, err := h.store.ListAllExternalMetricValues()
+	if err != nil {
+		log.Infof("Error while retrieving external metrics from the store: %s", err)
+		return
+	}
+
+	if len(emList) == 0 {
+		log.Debugf("No External Metrics to evaluate at the moment")
+		return
+	}
+
+	updated := h.hpaProc.UpdateExternalMetrics(emList)
+	if err = h.store.SetExternalMetricValues(updated); err != nil {
+		log.Errorf("Could not update the external metrics in the store: %s", err.Error())
+	}
+}
+
+// gc checks if any hpas have been deleted (possibly while the Datadog Cluster Agent was
+// not running) to clean the store.
+func (h *AutoscalersController) gc() {
+	log.Infof("Starting gc run")
+
+	list, err := h.clientSet.AutoscalingV2beta1().HorizontalPodAutoscalers(metav1.NamespaceAll).List(metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("Could not list hpas: %v", err)
+		return
+	}
+
+	emList, err := h.store.ListAllExternalMetricValues()
+	if err != nil {
+		log.Errorf("Could not list external metrics from store: %v", err)
+		return
+	}
+
+	deleted := h.hpaProc.ComputeDeleteExternalMetrics(list, emList)
+	if h.store.DeleteExternalMetricValues(deleted); err != nil {
+		log.Errorf("Could not delete the external metrics in the store: %v")
+		return
+	}
+	log.Debugf("Done GC run. Deleted %d metrics", len(deleted))
+}
+
+func (h *AutoscalersController) worker() {
+	for h.processNext() {
+	}
+}
+
+func (h *AutoscalersController) processNext() bool {
+	key, quit := h.queue.Get()
+	if quit {
+		return false
+	}
+
+	defer h.queue.Done(key)
+
+	if h.autoscalers != nil {
+		h.autoscalers <- key
+	}
+
+	ns, name, err := cache.SplitMetaNamespaceKey(key.(string))
+	if err != nil {
+		log.Errorf("Could not split the key: %v", err)
+		return false
+	}
+	hpa, err := h.autoscalersLister.HorizontalPodAutoscalers(ns).Get(name)
+	switch {
+	case errors.IsNotFound(err):
+		// We missed the Deletion Event. Local store does not have the HPA data anymore. The GC will clean up the Global Store.
+		log.Errorf("HorizontalPodAutoscaler %v has been deleted but was not caught in the EventHandler. GC will cleanup.", key)
+	case err != nil:
+		log.Errorf("Unable to retrieve Horizontal Pod Autoscaler %v from store: %v", key, err)
+	default:
+		if hpa == nil {
```

Comment:
Yes, I had a panic because the hpa had no metrics - I'd rather be extra cautious, although I know that this should be taken care of by the client-go's cache.